File: driveappfolderviewer/src/main/java/com/commit451/driveappfolderviewer/FilesAdapter.java
package com.commit451.driveappfolderviewer;

import android.support.v7.widget.RecyclerView;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupMenu;

import com.google.android.gms.drive.Metadata;

import java.util.ArrayList;

/**
 * Adapter which shows the files
 */
public class FilesAdapter extends RecyclerView.Adapter<FileViewHolder> {

    private ArrayList<Metadata> mMetadatas;
    private Listener mListener;

    private final View.OnClickListener mOnClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            Metadata metadata = (Metadata) view.getTag();
            mListener.onFileClicked(metadata);
        }
    };

    public FilesAdapter(Listener listener) {
        mMetadatas = new ArrayList<>();
        mListener = listener;
    }

    @Override
    public FileViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final FileViewHolder fileViewHolder = FileViewHolder.inflate(parent);
        fileViewHolder.itemView.setOnClickListener(mOnClickListener);
        fileViewHolder.popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem menuItem) {
                if (menuItem.getItemId() == R.id.action_delete) {
                    mListener.onDeleteClicked((Metadata) fileViewHolder.itemView.getTag());
                    return true;
                }
                if (menuItem.getItemId() == R.id.action_size) {
                    mListener.onSizeClicked((Metadata) fileViewHolder.itemView.getTag());
                    return true;
                }
                return false;
            }
        });
        return fileViewHolder;
    }

    @Override
    public void onBindViewHolder(FileViewHolder holder, int position) {
        Metadata metadata = mMetadatas.get(position);
        holder.itemView.setTag(metadata);
        holder.bind(metadata);
    }

    @Override
    public int getItemCount() {
        return mMetadatas.size();
    }

    public void setMetadatas(ArrayList<Metadata> metadatas) {
        mMetadatas.clear();
        mMetadatas.addAll(metadatas);
        notifyDataSetChanged();
    }

    public void clearMetadatas() {
        mMetadatas.clear();
        notifyDataSetChanged();
    }

    public interface Listener {
        void onDeleteClicked(Metadata metadata);
        void onFileClicked(Metadata metadata);
        void onSizeClicked(Metadata metadata);
    }
}

File: driveappfolderviewer/src/main/java/com/commit451/driveappfolderviewer/FilesAdapter.kt
package com.commit451.driveappfolderviewer

import android.view.ViewGroup
import android.widget.PopupMenu
import androidx.recyclerview.widget.RecyclerView
import com.google.api.services.drive.model.File

/**
 * Adapter which shows the files
 */
internal class FilesAdapter(private val listener: Listener) : RecyclerView.Adapter<FileViewHolder>() {

    private val files = mutableListOf<File>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FileViewHolder {
        val fileViewHolder = FileViewHolder.inflate(parent)
        fileViewHolder.itemView.setOnClickListener { view ->
            val metadata = view.tag as File
            listener.onFileClicked(metadata)
        }
        fileViewHolder.popupMenu.setOnMenuItemClickListener(PopupMenu.OnMenuItemClickListener { menuItem ->
            if (menuItem.itemId == R.id.action_delete) {
                listener.onDeleteClicked(fileViewHolder.itemView.tag as File)
                return@OnMenuItemClickListener true
            }
            if (menuItem.itemId == R.id.action_size) {
                listener.onSizeClicked(fileViewHolder.itemView.tag as File)
                return@OnMenuItemClickListener true
            }
            false
        })
        return fileViewHolder
    }

    override fun onBindViewHolder(holder: FileViewHolder, position: Int) {
        val metadata = files[position]
        holder.itemView.tag = metadata
        holder.bind(metadata)
    }

    override fun getItemCount(): Int {
        return files.size
    }

    fun setFiles(files: Collection<File>) {
        this.files.clear()
        this.files.addAll(files)
        notifyDataSetChanged()
    }

    fun clearFiles() {
        files.clear()
        notifyDataSetChanged()
    }

    interface Listener {
        fun onDeleteClicked(file: File)

        fun onFileClicked(file: File)

        fun onSizeClicked(file: File)
    }
}
